File: src/main/java/gargoyle/xenox/util/ini/PropertyMap.java
package gargoyle.xenox.util.ini;

import gargoyle.xenox.util.log.Log;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.nio.charset.Charset;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.function.Function;

@SuppressWarnings("WeakerAccess")
public final class PropertyMap {
    private final Map<String, String> entries = new HashMap<>();

    public PropertyMap(InputStream stream, Charset charset) {
        this(new InputStreamReader(stream, charset));
    }

    public PropertyMap(Reader reader) {
        try (BufferedReader bufferedReader = new BufferedReader(reader)) {
            load(bufferedReader);
        } catch (IOException e) {
            Log.error(e.getLocalizedMessage(), e);
        }
    }

    public Set<Map.Entry<String, String>> entrySet() {
        return entries.entrySet();
    }

    public String get(String key, String defaultValue) {
        return entries.getOrDefault(key, defaultValue);
    }

    public String get(String key) {
        return get(key, null);
    }

    public int get(String key, int defaultValue) {
        return Integer.parseInt(get(key, String.valueOf(defaultValue)));
    }

    public double get(String key, double defaultValue) {
        return Double.parseDouble(get(key, String.valueOf(defaultValue)));
    }

    public <T> T get(String key, T defaultValue, Function<String, T> parser) {
        return parser.apply(get(key, String.valueOf(defaultValue)));
    }

    private void load(BufferedReader reader) throws IOException {
        String line;
        while ((line = reader.readLine()) != null) {
            line = line.trim();
            if (line.isEmpty()) {
                continue;
            }
            if (line.startsWith("#") || line.startsWith("//")) {
                continue;
            }
            int i = line.indexOf('=');
            if (i > 0) {
                String key = line.substring(0, i).trim();
                String value = line.substring(i + 1).trim();
                put(key, value);
            }
        }
    }

    public void put(String key, String value) {
        entries.put(key, value);
    }
}

File: src/main/kotlin/gargoyle/xenox/util/ini/PropertyMap.kt
package gargoyle.xenox.util.ini

import gargoyle.xenox.util.log.Log
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.io.Reader
import java.nio.charset.Charset

class PropertyMap(reader: Reader) {
    private val entries: MutableMap<String, String?> = mutableMapOf()

    constructor(stream: InputStream, charset: Charset) : this(InputStreamReader(stream, charset))

    init {
        try {
            BufferedReader(reader).use(::load)
        } catch (e: IOException) {
            Log.error(e.localizedMessage, e)
        }
    }

    fun entrySet(): Set<Map.Entry<String, String?>> {
        return entries.entries
    }

    operator fun get(key: String, defaultValue: String? = null): String? {
        return entries.getOrDefault(key, defaultValue)
    }

    operator fun get(key: String, defaultValue: Int): Int {
        return get(key, defaultValue.toString())!!.toInt()
    }

    operator fun get(key: String, defaultValue: Double): Double {
        return get(key, defaultValue.toString())!!.toDouble()
    }

    operator fun <T> get(key: String, defaultValue: T, parser: (String) -> T): T {
        return parser(get(key, defaultValue.toString())!!)
    }

    private fun load(reader: BufferedReader) {
        var line0: String?
        while (reader.readLine().also { line0 = it } != null) {
            val line: String = line0!!.trim()
            if (line.isEmpty()) {
                continue
            }
            if (line.startsWith("#") || line.startsWith("//")) {
                continue
            }
            val i = line.indexOf('=')
            if (i > 0) {
                val key = line.substring(0, i).trim()
                val value = line.substring(i + 1).trim()
                put(key, value)
            }
        }
    }

    fun put(key: String, value: String?) {
        entries[key] = value
    }
}
